### ISSUE TITLE
Create a las file with the library

### ISSUE THREAD

**Issue by jabrena** [mreutegg/laszip4j#3]
Hi,

with this library, is it possible to create a las file from a data structure?

Many thanks in advance

Juan Antonio

**Comment by mreutegg** [comment]
No, this is currently not possible. This library is only a partial port of the original laszip library. So far the main focus was on reading existing .las and compressed .laz files.

**Comment by jabrena** [comment]
If you like, I could help in that issue. I am thinking to create .las / .laz files. 
If you are interested, I could do some PR.

**Comment by mreutegg** [comment]
Sure, feel free to give it a try and submit a PR. There are some classes like LASwriterLAS that only exist as skeletons. I didn't port them because so far I was primarily interested in reading .las and .laz files.

**Comment by jabrena** [comment]
Any las example to try to recreate?
I have a RPLidar A1 and I could create a .las/.laz  file.

**Comment by mreutegg** [comment]
Unfortunately the original C++ sources did not have any tests that allowed me to easily validate the ported code. You could try round-tripping existing data, e.g. referenced on [laszip.org](http://www.laszip.org/). The LAStools project also has a few [example files](https://github.com/LAStools/LAStools/tree/master/data).

**Comment by jabrena** [comment]
Hi @mreutegg,

It is a nice idea. Give me a time, I am debugging a ROS issue, when I finish, we will return to this conversation. We will collaborate.

Cheers

**Comment by AlexandreBrown** [comment]
Hello @jabrena @mreutegg , we are using the library to read las files and would like to write the las file with new classification values.  
Do you think this feature could be implemented in the future?  
Cheers

**Comment by mreutegg** [comment]
Hi Alexandre, I currently don't have plans to work on this feature. This project has a skeleton class [LASwriterLAS](https://github.com/mreutegg/laszip4j/blob/master/src/main/java/com/github/mreutegg/laszip4j/laslib/LASwriterLAS.java) you would have to implement or port from [LAStools/laswriter_las.cpp](https://github.com/LAStools/LAStools/blob/master/LASlib/src/laswriter_las.cpp).

**Comment by jabrena** [comment]
From my side, I am not working in robotics currently.

**Comment by mreutegg** [comment]
@AlexandreBrown, I spent some time implementing/porting the las writer.

See PR https://github.com/mreutegg/laszip4j/pull/73

It's only a start, but reading laz and writing the data into a las file seems to work.

**Comment by AlexandreBrown** [comment]
@mreutegg Thank you so much for you work.  
Here is how we read a LAS file currently:  
![image](https://user-images.githubusercontent.com/26939775/224562694-eb0fd41a-5761-48a1-b103-cd075968f2fe.png)
  
For writing, our use case is we need to write the `classification` field of a LAS file with new values for each rows.  
We also might need to support adding a new column/field to the LAS (to keep the existing LAS columns untouched).  
  
Let me know if you have any questions!

**Comment by mreutegg** [comment]
Hi @AlexandreBrown

Please have a look at https://github.com/mreutegg/laszip4j/pull/75

For now it just allows re-classification of points, but more modifier methods could be added easily. Does this go into the right direction?

> We also might need to support adding a new column/field to the LAS (to keep the existing LAS columns untouched).

Can you be more precise what you mean with column/field? AFAIK point data in LAS files have a fixed set of fields. I'm not aware that you can add more.

**Comment by AlexandreBrown** [comment]
@mreutegg Yes your PR seem to be exactly what we need in order to set the classification value.  
  
> Can you be more precise what you mean with column/field? AFAIK point data in LAS files have a fixed set of fields. I'm not aware that you can add more.  
  
Do you think it would be possible to add a new dimension to the LAS?  
Something similar to [this](https://pythonhosted.org//laspy/tut_part_3.html).  
  
In our case it would be used to add a new dimension that stores the new classification values while keeping the existing classification field untouched.

**Comment by mreutegg** [comment]
OK, I see. The extra dimension information is stored as 'extra bytes' in an extended variable length record (EVLRs). Currently laszip4j only has limited support for EVLRs. You can access them through the LASHeader, but the payload of a EVLR is simply exposed as a ByteBuffer backed by an in-memory byte array. This won't work well for large EVLRs. And then the question is how to write these.

Maybe EVLRs could be defined when a LASWriter is constructed and then populated via the transformer.

**Comment by AlexandreBrown** [comment]
@mreutegg I see.  
Since we're still unsure we'll 100% need this, perhaps we can close this issue since the "Create a las file with the library" is solved and improved EVLRs support can be added in an other issue.

**Comment by mreutegg** [comment]
Agreed. Though, I was wrong about the extra bytes in EVLRs. Only the metadata of the extra bytes are stored in EVLRs. The actual extra bytes are appended to the point data record. In the original laszip tool this is modeled as attributes on a LAS point. Some of that code has already been ported to laszip4j. But there's still quite some work to do to support writing of extra bytes. So, yes. Let's resolve this issue and create a new one if needed.
